client/search-ui/src/utilities/hooks.ts

```diff
@@ -0,0 +1,24 @@
+import React, {useState, useEffect, useRef, useCallback} from 'react';
+
+export function useStateCallback<T>(
+    initialState: T
+  ): [T, (state: T, cb?: (state: T) => void) => void] {
+    const [state, setState] = useState(initialState);
+    const cbRef = useRef<((state: T) => void) | undefined>(undefined); // init mutable ref container for callbacks
+  
+    const setStateCallback = useCallback((state: T, cb?: (state: T) => void) => {
+      cbRef.current = cb; // store current, passed callback in ref
+      setState(state);
+    }, []); // keep object reference stable, exactly like `useState`
+  
+    useEffect(() => {
+      // cb.current is `undefined` on initial render,
+      // so we only invoke callback on state *updates*
+      if (cbRef.current) {
+        cbRef.current(state);
+        cbRef.current = undefined; // reset callback after execution
+      }
+    }, [state]);
+  
+    return [state, setStateCallback];
```

Comment:
What is this for?
If you wanted a version of `useState` where you want the previous state, then the following use case should be enough:
```ts
const [state, setState] = useState(false);
setState((prev) => !prev);
```

Reply:
@Duckapple Sorry about that one ! It wasn't meant to be pushed ! 

So before React introduced hooks, you could put a call back when setting a state that would be called AFTER the state was modified. This is very useful when you want to be able to setState after another setState, or do anything that is dependent on the new state. 

Nowadays with hooks, you can have a similar thing by using the useEffect hook. However, this useEffect will always be called when the given arguments are modified, which includes even the initial render. This is very unpractical and can cause some havoc. Also it was really handy to be able to call a different callback contextually. 

This hook behaves similarly to the old class setState call back. For reference it comes from here: 
https://stackoverflow.com/questions/54954091/how-to-use-callback-with-usestate-hook-in-react

In our case I was thinking of using it for some functionalities such as saving the state in user workspace (vscode.saveState). 

However I will debug the panel which is currently wonky in a different branch and different PR. We can remove it from this PR.

Reply:
Ah, I see... I didn't know that behaviour existed back then, that's kinda neat!